intellij-toml/src/main/kotlin/org/toml/lang/CrateVersion.kt

```diff
@@ -0,0 +1,161 @@
+            val preBound = input.indexOf('+', preStartBound + 1).takeUnless { it == -1 } ?: input.length
+            val pres = if (preStartBound != -1 && preStartBound != preBound) input.substring(preStartBound + 1, preBound).split('.').map { Identifier.parse(it) } else listOf()
+
+            val builds = if (preBound + 1 < input.length) input.substring(preBound + 1).split('.').map { Identifier.parse(it) } else listOf()
+
+            return IncompleteCrateVersion(major, minor, patch, pres, builds)
+        }
+    }
+}
+
+data class CrateVersion(val major: Int, val minor: Int, val patch: Int, val pre: List<Identifier> = emptyList(), val build: List<Identifier> = emptyList()) : Comparable<CrateVersion> {
+    companion object {
+        /**
+         * Requires all 3 components (major, minor, patch) to be present.
+         */
+        fun parse(string: String): CrateVersion {
+            val input = string.trim()
+            val majorBound = input.indexOf('.').takeUnless { it == -1 }
+                ?: throw IllegalArgumentException("Major not provided")
+            val major = input.substring(0, majorBound).toInt()
+
+            val minorBound = input.indexOf('.', majorBound + 1).takeUnless { it == -1 }
+                ?: throw IllegalArgumentException("Minor not provided")
+            val minor = input.substring(majorBound + 1, minorBound).toInt()
+
+            val patchBound = input.indexOfAny(charArrayOf('-', '+'), minorBound + 1).takeUnless { it == -1 }
+                ?: input.length
+            val patch = input.substring(minorBound + 1, patchBound).toInt()
+
+            val preStartBound = if (input.getOrNull(patchBound) == '-') patchBound else -1
+            val preBound = input.indexOf('+', preStartBound + 1).takeUnless { it == -1 } ?: input.length
+            val pres = if (preStartBound != -1 && preStartBound != preBound) input.substring(preStartBound + 1, preBound).split('.').map { Identifier.parse(it) } else listOf()
+
+            val builds = if (preBound + 1 < input.length) input.substring(preBound + 1).split('.').map { Identifier.parse(it) } else listOf()
+
+            return CrateVersion(major, minor, patch, pres, builds)
+        }
+
+        fun tryParse(string: String): CrateVersion? = try {
+            parse(string)
+        } catch (e: IllegalArgumentException) {
+            null
+        }
+    }
+
+    fun isPrerelease() = pre.isNotEmpty()
+
+    override fun compareTo(other: CrateVersion): Int {
+        val majorCmp = this.major.compareTo(other.major)
+        if (majorCmp != 0) return majorCmp
+
+        val minorCmp = this.minor.compareTo(other.minor)
+        if (minorCmp != 0) return minorCmp
+
+        val patchCmp = this.patch.compareTo(other.patch)
+        if (patchCmp != 0) return patchCmp
+
+        val preCmp = this.pre.withIndex().firstOrNull { (i, own) -> other.pre.getOrNull(i)?.compareTo(own) != 0 }?.let {
```

Comment:
This code is very complex and hard to understand. Please try to reduce its complexity caused by:
* `firstOrNull(...)?.let(...) ?: ...`
* `times(-1)`
* `other.pre.getOrNull(preI)` usage